core/src/main/java/com/linecorp/armeria/internal/ArmeriaHttpUtil.java

```diff
@@ -319,34 +339,22 @@ public static void toArmeria(io.netty.handler.codec.http.HttpHeaders inHeaders,
                 }
 
                 final String value = entry.getValue().toString();
+                // Cookies must be concatenated into a single octet string.
+                // https://tools.ietf.org/html/rfc7540#section-8.1.2.5
                 if (aName.contentEqualsIgnoreCase(HttpHeaderNames.COOKIE)) {
-                    // split up cookies to allow for better compression
-                    // https://tools.ietf.org/html/rfc7540#section-8.1.2.5
-                    int index = value.indexOf(';');
-                    if (index >= 0) {
-                        int start = 0;
-                        do {
-                            out.add(HttpHeaderNames.COOKIE, value.substring(start, index));
-                            // skip 2 characters "; " (see https://tools.ietf.org/html/rfc6265#section-4.2.1)
-                            index++;
-                            if (index != value.length() && value.charAt(index) == ' ') {
-                                index++;
-                            }
-                            start = index;
-                        } while (start < value.length() &&
-                                 (index = value.indexOf(';', start)) != -1);
-
-                        if (start < value.length()) {
-                            out.add(HttpHeaderNames.COOKIE, value.substring(start));
-                        }
-                    } else {
-                        out.add(HttpHeaderNames.COOKIE, value);
+                    if (cookieJoiner == null) {
+                        cookieJoiner = new StringJoiner(COOKIE_SEPARATOR);
                     }
+                    COOKIE_SPLITTER.split(value).forEach(cookieJoiner::add);
```

Comment:
Shouldn't this conversion be in `toAmeria(Http2Headers` and not here? AFAIU, the spec is only related to HTTP/2 while a similar conversion is not needed for HTTP/1.

Reply:
Right. I added it just to make it more robust so it handles a poorly written peer. Should we just accept only the first one?

Reply:
.. and, yes, the conversion should also be in `toArmeria(Http2Headers)`. Added there as well.

Reply:
As a user I have some expectation that mutiple cookie headers would not be handled specially for HTTP/1 but think either way works. I'd just follow whatever tomcat etc does not sure what that is though :)